### ISSUE TITLE
KOReader on development build fails to connect to Calibre, "uuid" error

### ISSUE THREAD

**Issue by lightmaster** [koreader/koreader#12062]
* KOReader version: 2024.04-197
* Device: Kobo Libra 2 and Kobo Libra Colour

#### Issue

On the current development build (2024.04-197), KOReader cannot successfully connect to Calibre. On the current official build, it is able to without issue.

On my Kobo Libra Colour, on development build, when I try to wirelessly connect to Calibre, I get the following error message in Calibre:

```
calibre, version 7.12.0
ERROR: Error: Error communicating with device

'uuid'

Traceback (most recent call last):
  File "calibre\gui2\device.py", line 111, in run
  File "calibre\gui2\device.py", line 670, in _delete_books
  File "calibre\devices\smart_device_app\driver.py", line 57, in _synchronizer
  File "calibre\devices\smart_device_app\driver.py", line 1544, in delete_books
KeyError: 'uuid'
```

If I click "Ok", then Calibre will not be able to access the list of books in KOReader. If I move my books folder from `.Books` to `.Books.bak` and start a fresh `.Books` folder, then KOReader can connect to calibre successfully once. After uploading any number of books, disconnecting and reconnecting to Calibre, the error message is back again.

I tried it on my Kobo Libra 2 on the current official build (2024.04) since the Libra Colour will not work with the official build yet. KOReader on 2024.04 is able to connect to Calibre without any issue. When I then updated my Libra 2 from 2024.04 to 2024.04-197, it started getting the same error message.

This bug was also reported to the Calibre bug tracker by someone else: https://bugs.launchpad.net/calibre/+bug/2069006

#### Steps to reproduce

1. Install the current development build of KOReader (2024.04-197)
2. Start with an empty inbox folder in KOReader
3. Wirelessly connect to Calibre (no error on connection)
4. Send a book to KOReader and see the book show up in KOReader.
5. Disconnect from Calibre
6. Connect back to Calibre from KOReader.
7. Get "uuid" error message in Calibre

##### `crash.log` (if applicable)

[crash.log.zip](https://github.com/user-attachments/files/15926778/crash.log.zip)

**Comment by pazos** [comment]
Hi. Indeed we changed some things recently.

In the calibre plugin settings, on nightlies, you'll see a JSON parser submenu. It should have "automatic" checked. You'll need to test the other options.

But first and foremost, start with a clean state. There's a chance the JSON file is actually corrupted. To do that you can setup the inbox dir to a different directory.

With a clean inbox dir select the "legacy" json parser. Start a connection. Push some books, unconnect & reconnect. Verify everything works. Then switch to "safe" and do the same. If, at that point, everything works, switch again to automatic/fast and verify if you can reproduce the issue.

**Comment by NiLuJe** [comment]
[Oopsie](https://github.com/NiLuJe/lua-rapidjson/commit/e84973356255bde06a70ce6263a3a0ef5c8f4ad4) ;p.

**Comment by Frenzie** [comment]
Well, I'm glad ya'll introduced some release blockers and it wasn't just me being sick delaying things. ;-P

**Comment by lightmaster** [comment]
> But first and foremost, start with a clean state. There's a chance the JSON file is actually corrupted. To do that you can setup the inbox dir to a different directory.
> 
> With a clean inbox dir select the "legacy" json parser. Start a connection. Push some books, unconnect & reconnect. Verify everything works. Then switch to "safe" and do the same. If, at that point, everything works, switch again to automatic/fast and verify if you can reproduce the issue.

Already tried that. With a fresh install of KOReader (only setting changed is Calibre IP/password) and a blank books folder, it lets me connect the first time and send books over, but on the second connection the error is back.

Also, it might be corrupting the metadata file as on my Libra 2, once I downgraded back to official build, I couldn't use wireless Calibre still. Need to test on that one if a fresh books folder allows it to work.

**Comment by lightmaster** [comment]
Just saw @NiLuJe's message. I'll try with the next nightly to see if that fixed it.

**Comment by kyxap** [comment]
btw @lightmaster how do you do wireless connection, which app/plugin/tool are you using? just curios

**Comment by lightmaster** [comment]
> btw @lightmaster how do you do wireless connection, which app/plugin/tool are you using? just curios

Not sure what you mean... KOReader has a built in Calibre plugin and it connects to Calibre running on my computer.

**Comment by kyxap** [comment]
I've tested with `koreader-kobo-v2024.04-197-gb5a822cf8_2024-06-20.zip` on empty .metadata file:
- legacy (upload two books, no issue after reconnect)
- safe  (upload two books, no issue after reconnect)
- automatic:
<img width="400" alt="image" src="https://github.com/koreader/koreader/assets/3080529/5cadb056-e7f4-4ba1-b84f-2ce7c03c201e">
(do not look into calibre code line numbers, it's latest build with some debug enabled)

so kinda not sure what does it tell, its not fixed? or change did not go into that build?

**Comment by benoit-pierre** [comment]
It's not in master yet.

**Comment by kyxap** [comment]
kinda off top but related is there a way to force kobo os to rescan device and update sqlite  (aka book import)? Because it seems like it  works randomly if do exist/reboot/disconnect/or else, actually I do not know what gonna trigger that action if koreader was loaded.

Also I see that something (koreader?) deletes `uuid` from `metadata.calibre` after I do wired upload and calibre reopen while koreader is not running. 
Hard to find the way to use calibre with koreader if I decide to connect and seems calibre pull data from sqlite to match device book with calibre db

**Comment by NiLuJe** [comment]
> is there a way to force kobo os to rescan device

NickelMenu.

e.g.,
```
menu_item : library : Rescan      : nickel_misc : rescan_books
menu_item : library : Full Rescan : nickel_misc : rescan_books_full
```

(You almost never need the full variant).

>  I do wired upload and calibre reopen while koreader is not running.

You can't mix a wireless inbox with a USB library. There's a giant warning shown when you do it the other way around.

**Comment by kyxap** [comment]
>  There's a giant warning shown when you do it the other way around (i.e., attempt to set your wireless inbox to a kobo driver library).

I have not seen any warnings, is this should be somewhere in calibre or koreader? (I have subfolder for books and main calibre metadata saves in root)

**Comment by NiLuJe** [comment]
Us, of course, I've just said we obviously can't do anything about it the other way around ;).

https://github.com/koreader/koreader/blob/799275d3a3b054926a40d42f2b7dc1e1db990e14/plugins/calibre.koplugin/wireless.lua#L208-L216

**Comment by kyxap** [comment]
> Us, of course, I've just said we obviously can't do anything about it the other way around ;).
> 
> https://github.com/koreader/koreader/blob/799275d3a3b054926a40d42f2b7dc1e1db990e14/plugins/calibre.koplugin/wireless.lua#L208-L216

never happened to me. not sure what is 'folder' there

**Comment by lightmaster** [comment]
@kyxap What I did was disable the kobo usb driver in Calibre and set my books folder in KOReader to ".Books" on the SSD. This way the Kobo OS doesn't see the books at all. When I connect KOReader via USB, it gets mounted as a USB Drive, and I go to "Connect/share" on Calibre's menu bar and then to "Connect to folder" and select the ".Books" folder (might need to enable seeing hidden files in your OS and in KOReader's settings). For both USB and Wireless devices, I have Calibre configured to use the following "Save template" (use whatever you want, just make sure they are the same on both device types): `{series:ifempty({author_sort})}/{series_index:|| - }{title} - {author_sort}`. This places the books in folders for the series if it's part of a series, else it places them in Author folder, and if it is part of a series, the book filename then starts with the series index number to ensure they are always sorted in the correct order.

@NiLuJe I assume that since this was marked as complete, it should show up in the next nightly build? The [normal nightly url](https://build.koreader.rocks/download/nightly/) isn't working for me right now, but I did find [this link](https://gitlab.com/koreader/nightly-builds/-/pipelines) that seems to be working.
